sound/soc/intel/avs/pcm.c

```diff
@@ -278,12 +278,33 @@ static const struct snd_soc_dai_ops avs_dai_hda_be_ops;
 
 static int avs_dai_hda_be_startup(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
 {
-	return avs_dai_startup(substream, dai, false, &avs_dai_hda_be_ops);
+	struct avs_dev *adev = to_avs_dev(dai->dev);
+	struct hdac_bus *bus = &adev->base.core;
+	struct hdac_ext_stream *link_stream;
+	int ret;
+
+	link_stream = snd_hdac_ext_stream_assign(bus, substream, HDAC_EXT_STREAM_TYPE_LINK);
+	if (!link_stream)
+		return -EBUSY;
+
+	ret = avs_dai_startup(substream, dai, false, &avs_dai_hda_be_ops);
+	if (ret)
+		snd_hdac_ext_stream_release(link_stream, HDAC_EXT_STREAM_TYPE_LINK);
+	else
+		substream->private_data = link_stream;
+
+	return ret;
 }
 
 static void avs_dai_hda_be_shutdown(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
 {
-	return avs_dai_nonhda_be_shutdown(substream, dai);
+	struct hdac_ext_stream *link_stream;
+
+	link_stream = snd_substream_to_hext_stream(substream);
+
+	avs_dai_nonhda_be_shutdown(substream, dai);
+	snd_hdac_ext_stream_release(link_stream, HDAC_EXT_STREAM_TYPE_LINK);
+	substream->private_data = NULL;
```

Comment:
I remember asking @ranj063 why we currently assign the link DMA in hw_params, and of course I don't recall the answer.

Reply:
I believe that's because how the skylake-driver does things. LINK streams assigned in `hw_params()`, HOST streams in `open()`.